test/object-store/sync/flx_role_change.cpp

```diff
@@ -0,0 +1,861 @@
+            REQUIRE(config_1.sync_config->user->user_id() != config_2.sync_config->user->user_id());
+            std::atomic<bool> test_started = false;
+
+            // Reopen realm 2 and add a hook callback to check for bootstraps, which should not happen
+            // on this realm
+            config_2.sync_config->on_sync_client_event_hook = [&](std::weak_ptr<SyncSession>,
+                                                                  const SyncClientHookData& data) {
+                using Event = SyncClientHookEvent;
+                if (!test_started.load()) {
+                    return SyncClientHookAction::NoAction; // Not checking yet
+                }
+                // If a download message was received or bootstrap was processed, then fail the test
+                if ((data.event == Event::DownloadMessageReceived &&
+                     data.batch_state != sync::DownloadBatchState::SteadyState) ||
+                    data.event == Event::BootstrapMessageProcessed || data.event == Event::BootstrapProcessed) {
+                    FAIL("Bootstrap occurred on the second realm, which was not expected");
+                }
+                return SyncClientHookAction::NoAction;
+            };
+            auto realm_2 = Realm::get_shared_realm(config_2);
+            set_up_realm(realm_2, params.num_emps);
+
+            test_started = true;
+            // The first rule allows access to all records for user 1
+            AppCreateConfig::ServiceRole user1_role{"user 1 role"};
+            user1_role.apply_when = {{"%%user.id", config_1.sync_config->user->user_id()}};
+            // Add two rules, the first applies to user 1 and the second applies to other users
+            test_rules["roles"] = {transform_service_role(user1_role), transform_service_role(general_role)};
+            update_perms_and_verify(*harness, realm_1, test_rules,
+                                    {BootstrapMode::Any, params.num_emps, params.num_mgrs, params.num_dirs});
+
+            // Realm 2 data should not change (and there shouldn't be any bootstrap messages)
+            verify_records(realm_2, params.num_emps, 0, 0);
+
+            // The first rule will be updated to only have access to employee and managers
+            AppCreateConfig::ServiceRole user1_role_2 = user1_role;
+            user1_role_2.document_filters.read = {{"role", {{"$in", {"employee", "manager"}}}}};
+            user1_role_2.document_filters.write = {{"role", {{"$in", {"employee", "manager"}}}}};
+            // Update the first rule for user 1 and verify the data after the rule is applied
+            test_rules["roles"][0] = {transform_service_role(user1_role_2)};
+            update_perms_and_verify(*harness, realm_1, test_rules,
+                                    {BootstrapMode::Any, params.num_emps, params.num_mgrs, 0});
+
+            // Realm 2 data should not change (and there shouldn't be any bootstrap messages)
```

Comment:
If there shouldn't be any bootstrap messages, why do we pass BootstrapMode::Any to update_perms_and_verify above?

Reply:
I added a comment for this - realm_1 should receive a role change bootstrap, but realm_2 should not (and there is a `on_sync_client_event_hook` fcn for realm_2 that verifies this).